### DISCUSSION TITLE
Optimize CPU and GPU usage

### DISCUSSION THREAD

**Discussion by doczoidberg** [handsontable/hyperformula#566] [General]
we have a sheet which takes 25 seconds to set the content. But either CPU (~10%) not GPU (0%) are used. There should be room for improvements?

Is there a way how a sheet should be created to use the gpu?

**Comment by bardek8** [comment]
Well, this is strange that we do not use your CPU in full. Do you mean a single core or 20% altogether?
Have you tried profiling that? If not, is it the sheet you've already provided me or some other? Maybe we can take a look a the performance.

GPU improves performance in only some limited cases, with heavy operations on matrices. See the [note](https://handsontable.github.io/hyperformula/guide/performance.html#gpu-acceleration) in the documentation. I doubt if this would be applicable to your case (at least for the examples I've seen).

**Comment by doczoidberg** [comment]
I sent you a mail with the json file.
![image](https://user-images.githubusercontent.com/1972209/99006012-18424e80-2542-11eb-8eaa-0f9aba40fd7c.png)


Setting the contents also freezes my browser tab btw. Is there a possibility to run this in background?

**Comment by bardek8** [comment]
So there are few points here, let me summarize: 
- in principle, due to the usual sequential structure of spreadsheets, HF does not speed up well with the larger number of CPU cores. If you run it on a single core, you would get 100% CPU usage. Notice that at the right half of the charts you have ~40% usage of your all CPUs 
- as I mentioned, GPU improves performance only in some limited cases (with heavy operations on matrices), so it will not help in your case
- we believe it will be possible to run HF in a separate web worker ("in background"), so that the app does not freeze. However, due to the variety of applications, we don't have a reference implementation available.
- in, general your spreadsheet is really big, with ~10^6 non-empty cells, so this is quite a challenge for us :)

Looking at your example we have some general recommendations, how to improve performance of HF only by slightly fixing the spreadsheet:
- try to avoid including empty/unused cells in the ranges if they are irrelevant. For example, even if you know that there are no cells below the 100-th row and `SUM(A1:A100)` gives the same result as `SUM(A1:A200)`, we need to inspect the other cells and be prepared that they might change in the future,
- in particular, references to the whole columns (e.g. `SUM(A:B)/VLOOKUP(A:B,) are very costly and avoid them if possible.
- try to rearrange columns for `VLOOKUP`, so that there are not irrelevant cells in-between. For example in `VLOOKUP("abcd",B1:U10,8)` one may think that there are only 20 cells that should be considered here, but in the current implementation we don't optimize it and we will process all the 200 cells.

In short, HF will benefit most from rewriting this kind of formulas: `VLOOKUP("abcd",B:U,8)`.

We understand that this is not always possible, so probably in the future we'll consider how to optimize these case automatically.

**Reply by MartinDawson** [reply]
@bardek8 

> we believe it will be possible to run HF in a separate web worker ("in background"), so that the app does not freeze. However, due to the variety of applications, we don't have a reference implementation available.

Does this mean hyperformula has a plan to do this in the future or we should just do it now on our side?

**Reply by wojciechczerniak** [reply]
@MartinDawson No, we don't plan such a feature. There are multiple ways to use threading, Worker, SharedWorker, ServiceWorker, server-side child process on node. It depends on the final implementation and should be left for the end developer to decide

**Comment by doczoidberg** [comment]
thank you for the information @bardek8 . I think we can optimize the sheets on our side
